### PR TITLE
convert rgbd files to pcd (pointcloud)

### PR DESCRIPTION
**Pull request by PetervDooren**

### CODE REVIEW

File: rgbd/tools/rgbd_to_pcd.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        {
+            for (int j = 0; j < cloud.width; j++)
+            {
+                double x = image.getDepthImage().at<float>(i,j);
+                double y = (half_height-i) * x / fy;
+                double z = (-half_width+j) * x / fx;
+
+                cv::Vec3b pixel = image.getRGBImage().at<cv::Vec3b>(i,j);
+                uint b = pixel[0]; uint g = pixel[1];
+                uint r = pixel[2];
```
Comment (MatthijsBurgh) [inline rgbd/tools/rgbd_to_pcd.cpp:68]:
```suggestion
                uint b = pixel[0];
                uint g = pixel[1];
                uint r = pixel[2];
```

File: rgbd/tools/rgbd_to_pcd.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        cloud.resize (cloud.width * cloud.height);
+
+        double fx = image.getCameraModel().fx();
+        double fy = image.getCameraModel().fy();
+
+        double half_height = 0.5 * cloud.height;
+        double half_width = 0.5 * cloud.width;
+        for (int i = 0; i < cloud.height; i++)
+        {
+            for (int j = 0; j < cloud.width; j++)
```
Comment (MatthijsBurgh) [inline rgbd/tools/rgbd_to_pcd.cpp:60]:
```suggestion
        for (uint i = 0; i < cloud.height; ++i)
        {
            for (uint j = 0; j < cloud.width; ++j)
```

File: rgbd/tools/rgbd_to_pcd.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        tue::serialization::InputArchive a_in(f_in);
+
+        rgbd::Image image;
+        rgbd::deserialize(a_in, image);
+
+        pcl::PointCloud<pcl::PointXYZRGB> cloud;
+
+        // Fill in the cloud data
+        cloud.width    = image.getRGBImage().cols;
+        cloud.height   = image.getRGBImage().rows;
```
Comment (MatthijsBurgh) [inline rgbd/tools/rgbd_to_pcd.cpp:49]:
```suggestion
        cloud.width = image.getRGBImage().cols;
        cloud.height = image.getRGBImage().rows;
```

File: rgbd/tools/rgbd_to_pcd.cpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        cloud.is_dense = false;
+        cloud.resize (cloud.width * cloud.height);
+
+        double fx = image.getCameraModel().fx();
+        double fy = image.getCameraModel().fy();
+
+        double half_height = 0.5 * cloud.height;
+        double half_width = 0.5 * cloud.width;
+        for (uint i = 0; i < cloud.height; ++i)
```
Comment (MatthijsBurgh) [inline rgbd/tools/rgbd_to_pcd.cpp:58]:
```suggestion
        for (uint i=0; i < cloud.height; ++i)
```

File: rgbd/tools/rgbd_to_pcd.cpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        {
+            for (uint j = 0; j < cloud.width; ++j)
```
Comment (MatthijsBurgh) [inline rgbd/tools/rgbd_to_pcd.cpp:60]:
```suggestion
            for (uint j=0; j < cloud.width; ++j)
```

File: rgbd/tools/rgbd_to_pcd.cpp [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+#include <opencv2/highgui/highgui.hpp>
+
+#include <typeinfo>
+
+
+int main(int argc, char **argv) {
+
+    if (argc < 2)
+    {
+        std::cout << "Usage:\n\n   rgbd_to_pcd FILENAME.rgbd\n\n";
```
Comment (MatthijsBurgh) [inline rgbd/tools/rgbd_to_pcd.cpp:22]: Make it clear that you can provide multiple rgbd files.

File: rgbd/tools/rgbd_to_pcd.cpp [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        return 1;
+    }
+
+    for (int i = 1; i < argc; ++i)
```
Comment (MatthijsBurgh) [inline rgbd/tools/rgbd_to_pcd.cpp:26]:
```suggestion
    for (uint i=1; i < argc; ++i)
```

File: rgbd/tools/rgbd_to_pcd.cpp [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+                uint b = pixel[0];
+                uint g = pixel[1];
+                uint r = pixel[2];
+
+                cloud.at(j,i).x = x;
+                cloud.at(j,i).y = y;
+                cloud.at(j,i).z = z;
+                cloud.at(j,i).r = r;
+                cloud.at(j,i).g = g;
+                cloud.at(j,i).b = b;
```
Comment (MatthijsBurgh) [inline rgbd/tools/rgbd_to_pcd.cpp:76]: You could reduce the memory by directly writing to the cloud.

File: rgbd/tools/rgbd_to_pcd.cpp [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+            }
+        }
+
+        size_t lastindex = name.find_last_of(".");
+        name = name.substr(0, lastindex);
+
+        std::string pcd_filename = name + ".pcd";
+
+        pcl::io::savePCDFileASCII (pcd_filename, cloud);
+        std::cerr << "Saved " << cloud.size () << " data points to " << pcd_filename << std::endl;
```
Comment (MatthijsBurgh) [inline rgbd/tools/rgbd_to_pcd.cpp:86]: I think this should be `std::cout`

**Review by MatthijsBurgh — CHANGES_REQUESTED** [review]
Some minor changes. Maybe add some more comments to make the code easier to read.

File: rgbd/tools/rgbd_to_pcd.cpp [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+#include <opencv2/highgui/highgui.hpp>
+
+#include <typeinfo>
+
+
+int main(int argc, char **argv) {
+
+    if (argc < 2)
+    {
+        std::cout << "Usage:\n\n   rgbd_to_pcd FILENAME.rgbd (optional)SECOND_FILENAME.rgbd\n\n";
```
Comment (MatthijsBurgh) [inline rgbd/tools/rgbd_to_pcd.cpp:22]:
```suggestion
        std::cout << "Usage:\n\n   rgbd_to_pcd FILENAME.rgbd [FILENAME2.rgbd ...]\n\n";
```
This is a more common way to represent the optional extra arguments

**Review by MatthijsBurgh — CHANGES_REQUESTED** [review]
One small minor change

File: rgbd/tools/rgbd_to_pcd.cpp (hunk 10 above)
Comment (PetervDooren) [reply rgbd/tools/rgbd_to_pcd.cpp:22]: Thanks

**Review by MatthijsBurgh — APPROVED** [review]

**Pull request merged**